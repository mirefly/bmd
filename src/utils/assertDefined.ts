export default function assertDefined<T>(
  value: T | null | undefined,
  message = 'value expected to be defined but was not'
): T {
  if (value === null || value === undefined) {
    throw new Error(message)
  }

  return value
}
