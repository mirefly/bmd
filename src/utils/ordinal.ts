/**
 * Returns number param as ordinal.
 * Source: https://stackoverflow.com/a/31615643/101290
 *
 * @param n number
 *
 * @return ordinal string
 *
 */

function ordinal(n: number): string {
  const s = ['th', 'st', 'nd', 'rd']
  const v = n % 100
  return n + (s[(v - 20) % 10] || s[v] || s[0])
}
export default ordinal
