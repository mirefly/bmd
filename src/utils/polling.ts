export interface Poller {
  stop(): void
}

export class IntervalPoller implements Poller {
  private duration: number
  private callback: () => Promise<void> | void
  private interval?: ReturnType<typeof setTimeout>

  public constructor(duration: number, callback: () => Promise<void> | void) {
    this.duration = duration
    this.callback = callback
  }

  public start(): this {
    if (!this.interval) {
      this.pollForever()
    }

    return this
  }

  public stop(): this {
    window.clearInterval(this.interval)
    this.interval = undefined
    return this
  }

  private pollForever(): void {
    this.interval = window.setInterval(async () => {
      try {
        await this.callback()
      } catch (error) {
        // ignore errors for now
      }
    }, this.duration)
  }

  public static start(
    interval: number,
    callback: () => Promise<void> | void
  ): IntervalPoller {
    return new IntervalPoller(interval, callback).start()
  }
}
