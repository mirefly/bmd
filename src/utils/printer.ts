import fetchJSON from './fetchJSON'

interface StatusResponse {
  ok: boolean
}

interface NewJobResponse {
  id: string
}

export interface PrintJob {
  id: string
  owner: Printer
}

export enum PrintMethod {
  LocalOnly = 'LocalOnly',
  RemoteOnly = 'RemoteOnly',
  RemoteWithLocalFallback = 'RemoteWithLocalFallback',
}

export enum PrintStatus {
  Pending = 'Pending',
  Printing = 'Printing',
  Success = 'Success',
  Error = 'Error',
  Unknown = 'Unknown',
}

export interface Printer {
  isReady(): Promise<boolean>
  print(): Promise<PrintJob>
  getStatus(job: PrintJob): Promise<PrintStatus>
}

export class LocalPrinter implements Printer {
  public async isReady(): Promise<boolean> {
    return true
  }

  public async print(): Promise<PrintJob> {
    window.print()
    return { id: '__local', owner: this }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public async getStatus(job: PrintJob): Promise<PrintStatus> {
    return PrintStatus.Unknown
  }
}

export class NullPrinter implements Printer {
  public async isReady(): Promise<boolean> {
    return true
  }

  public async print(): Promise<PrintJob> {
    return { id: '__nothing', owner: this }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public async getStatus(job: PrintJob): Promise<PrintStatus> {
    return PrintStatus.Unknown
  }
}

export class RemotePrinter implements Printer {
  public async isReady(): Promise<boolean> {
    try {
      const status = await fetchJSON<StatusResponse>('/printer/status')

      return status.ok
    } catch {
      return false
    }
  }

  public async print(): Promise<PrintJob> {
    if (!(await this.isReady())) {
      throw new Error('unable to print: remote printer is not ready')
    }

    const newJob = await fetchJSON<NewJobResponse>('/printer/jobs/new', {
      method: 'POST',
      headers: {
        'Content-Type': 'text/html',
      },
      body: document.documentElement.outerHTML,
    })

    return { id: newJob.id, owner: this }
  }

  /**
   * @todo implement this when the server implements `/printer/jobs/{id}`
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public async getStatus(_job: PrintJob): Promise<PrintStatus> {
    return PrintStatus.Unknown
  }
}

export class FallbackPrinters implements Printer {
  private readonly printers: Printer[]

  public constructor(printers: Printer[]) {
    if (printers.length === 0) {
      throw new Error('at least one printer is required')
    }

    this.printers = printers
  }

  public async isReady(): Promise<boolean> {
    for (const printer of this.printers) {
      if (await printer.isReady()) {
        return true
      }
    }

    return false
  }

  public async print(): Promise<PrintJob> {
    for (const printer of this.printers) {
      if (await printer.isReady()) {
        return printer.print()
      }
    }

    throw new Error('no printers were ready to print')
  }

  public async getStatus(job: PrintJob): Promise<PrintStatus> {
    for (const printer of this.printers) {
      if (printer === job.owner) {
        return printer.getStatus(job)
      }
    }

    throw new Error(`no printers own job: ${job.id}`)
  }
}

export default function makePrinter(method: PrintMethod): Printer {
  switch (method) {
    case PrintMethod.LocalOnly:
      return new LocalPrinter()

    case PrintMethod.RemoteOnly:
      return new RemotePrinter()

    case PrintMethod.RemoteWithLocalFallback:
      return new FallbackPrinters([new RemotePrinter(), new LocalPrinter()])

    default:
      throw new TypeError(`unknown print method: ${method}`)
  }
}
