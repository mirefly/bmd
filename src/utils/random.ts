/**
 * Random Base64 string.
 * Get random values to fill an array of prescribed length,
 * b64 encode it, and strip the '=' at the end.
 *
 * @param {number} [numBytes=16] Number of bytes
 *
 * @return string
 *
 */

export function randomBase64(numBytes = 16) {
  return window
    .btoa(
      String.fromCharCode(
        ...window.crypto.getRandomValues(new Uint8ClampedArray(numBytes))
      )
    )
    .replace(/=+$/, '')
}

export default { randomBase64 }
