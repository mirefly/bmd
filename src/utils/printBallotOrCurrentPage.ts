import { CompletedBallot } from '@votingworks/ballot-encoder'
import ballotToDocument from '@votingworks/ballot-renderer/renderers/ballotToDocument'
import { Printer, PrintType } from './printer'

/**
 * Prints the ballot to PDF if possible, otherwise prints the current page.
 * This assumes that printing the current page will print the ballot, which must
 * be set up before this is called.
 */
export default async function printBallotOrCurrentPage(
  printer: Printer,
  ballot: CompletedBallot
): Promise<void> {
  if (await printer.canPrint(PrintType.PDFMakeDocument)) {
    const document = ballotToDocument(ballot)

    await printer.print({ type: PrintType.PDFMakeDocument, pdf: document })
  } else {
    await printer.print({ type: PrintType.CurrentPage })
  }
}
