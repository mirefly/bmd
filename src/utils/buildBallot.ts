import {
  BallotType,
  CompletedBallot,
  Election,
  getBallotStyle,
  getPrecinctById,
  VotesDict,
} from '@votingworks/ballot-encoder'
import assertDefined from './assertDefined'
import { randomBase64 } from './random'

/**
 * Builds a ballot from the app state.
 */
export default function buildBallot({
  ballotStyleId,
  election,
  isTestBallot,
  precinctId,
  votes,
}: {
  ballotStyleId: string
  election: Election
  isTestBallot: boolean
  precinctId: string
  votes: VotesDict
}): CompletedBallot {
  const ballotId = randomBase64()
  const ballotStyle = assertDefined(getBallotStyle({ ballotStyleId, election }))
  const ballotType = BallotType.Standard
  const precinct = assertDefined(getPrecinctById({ precinctId, election }))

  return {
    ballotId,
    ballotStyle,
    ballotType,
    election,
    isTestBallot,
    precinct,
    votes,
  }
}
