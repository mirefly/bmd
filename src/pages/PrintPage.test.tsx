import React from 'react'
import { Route } from 'react-router-dom'

import {
  vote,
  electionSample,
  getContests,
  getBallotStyle,
} from 'ballot-encoder'

import { render, mockOf } from '../../test/testUtils'

import electionSampleWithSeal from '../data/electionSampleWithSeal.json'
import electionSampleNoSeal from '../data/electionSampleNoSeal.json'

import PrintPage from './PrintPage'

import { randomBase64 } from '../utils/random'

// mock the random value so the snapshots match
jest.mock('../utils/random')
const randomBase64Mock = mockOf(randomBase64)
randomBase64Mock.mockReturnValue('CHhgYxfN5GeqnK8KaVOt1w')

it('renders PrintPage without votes', () => {
  const { container } = render(<Route path="/print" component={PrintPage} />, {
    ballotStyleId: '5',
    precinctId: '21',
    route: '/print',
  })
  expect(container.firstChild).toMatchSnapshot()
})

it('renders PrintPage with votes', () => {
  const { container } = render(<Route path="/print" component={PrintPage} />, {
    ballotStyleId: '5',
    precinctId: '21',
    route: '/print',
    votes: vote(
      getContests({
        ballotStyle: getBallotStyle({
          election: electionSample,
          ballotStyleId: '5',
        })!,
        election: electionSample,
      }),
      {
        president: 'barchi-hallaren',
        'question-a': 'no',
        'question-b': 'yes',
        'lieutenant-governor': 'norberg',
      }
    ),
  })
  expect(container.childNodes[1]).toMatchSnapshot()
})

it('renders PrintPage without votes and inline seal', () => {
  const { container } = render(<Route path="/print" component={PrintPage} />, {
    ballotStyleId: '5',
    election: electionSampleWithSeal,
    precinctId: '21',
    route: '/print',
  })
  expect(container.childNodes[1]).toMatchSnapshot()
})

it('renders PrintPage without votes and no seal', () => {
  const { container } = render(<Route path="/print" component={PrintPage} />, {
    ballotStyleId: '5',
    election: electionSampleNoSeal,
    precinctId: '21',
    route: '/print',
  })
  expect(container.childNodes[1]).toMatchSnapshot()
})
