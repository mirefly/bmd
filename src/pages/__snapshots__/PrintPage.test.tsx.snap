// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PrintPage with votes 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c2 {
  margin: auto 0;
}

.c3 {
  margin: auto;
  max-width: 66ch;
  text-align: center;
  line-height: 1.2;
}

.c3 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c3 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c3 h3 {
  font-size: 1.17rem;
}

.c3 h4 {
  font-size: 1rem;
}

.c3 h5 {
  font-size: 0.9rem;
}

.c3 h3,
.c3 h4,
.c3 h5,
.c3 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p {
  margin-top: -0.75rem;
}

.c3 h3 + p,
.c3 h4 + p,
.c3 h5 + p {
  margin-top: -1rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c3 dl {
  margin: 1rem 0;
}

.c4 {
  max-width: 66ch;
  line-height: 1.2;
}

.c4 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c4 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c4 h3 {
  font-size: 1.17rem;
}

.c4 h4 {
  font-size: 1rem;
}

.c4 h5 {
  font-size: 0.9rem;
}

.c4 h3,
.c4 h4,
.c4 h5,
.c4 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c4 h1 + h2 {
  margin-top: -0.75rem;
}

.c4 h1 + p,
.c4 h2 + p {
  margin-top: -0.75rem;
}

.c4 h3 + p,
.c4 h4 + p,
.c4 h5 + p {
  margin-top: -1rem;
}

.c4 > :first-child {
  margin-top: 0;
}

.c4 > :last-child {
  margin-bottom: 0;
}

.c4 dl {
  margin: 1rem 0;
}

.c5 {
  margin-left: -1em;
  text-align: center;
  white-space: nowrap;
}

.c5::before,
.c5::after {
  display: inline-block;
  width: 0;
  overflow: hidden;
  vertical-align: bottom;
  text-align: left;
  content: '…';
  -webkit-animation: loadingEllipsis steps(4,end) 2s infinite;
  animation: loadingEllipsis steps(4,end) 2s infinite;
}

.c5::before {
  color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
}

.c0 > nav {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 > main {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

@media print {
  .c1 {
    display: none;
  }
}

@media print {
  .c2 {
    margin: 0;
    max-width: 100%;
  }
}

@media (min-width:480px) {
  .c3 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c4 {
    line-height: 1.3;
  }
}

@media print {
  .c0 {
    display: block;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
        id="audiofocus"
      >
        <h1
          aria-label="Printing Official Ballot."
        >
          <div
            class="c4"
          >
            <h1
              aria-label="Printing Official Ballot."
              class="c5"
            >
              Printing Official Ballot
            </h1>
          </div>
        </h1>
      </div>
    </div>
  </main>
</div>
`;

exports[`renders PrintPage without votes 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c2 {
  margin: auto 0;
}

.c3 {
  margin: auto;
  max-width: 66ch;
  text-align: center;
  line-height: 1.2;
}

.c3 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c3 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c3 h3 {
  font-size: 1.17rem;
}

.c3 h4 {
  font-size: 1rem;
}

.c3 h5 {
  font-size: 0.9rem;
}

.c3 h3,
.c3 h4,
.c3 h5,
.c3 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p {
  margin-top: -0.75rem;
}

.c3 h3 + p,
.c3 h4 + p,
.c3 h5 + p {
  margin-top: -1rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c3 dl {
  margin: 1rem 0;
}

.c4 {
  max-width: 66ch;
  line-height: 1.2;
}

.c4 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c4 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c4 h3 {
  font-size: 1.17rem;
}

.c4 h4 {
  font-size: 1rem;
}

.c4 h5 {
  font-size: 0.9rem;
}

.c4 h3,
.c4 h4,
.c4 h5,
.c4 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c4 h1 + h2 {
  margin-top: -0.75rem;
}

.c4 h1 + p,
.c4 h2 + p {
  margin-top: -0.75rem;
}

.c4 h3 + p,
.c4 h4 + p,
.c4 h5 + p {
  margin-top: -1rem;
}

.c4 > :first-child {
  margin-top: 0;
}

.c4 > :last-child {
  margin-bottom: 0;
}

.c4 dl {
  margin: 1rem 0;
}

.c5 {
  margin-left: -1em;
  text-align: center;
  white-space: nowrap;
}

.c5::before,
.c5::after {
  display: inline-block;
  width: 0;
  overflow: hidden;
  vertical-align: bottom;
  text-align: left;
  content: '…';
  -webkit-animation: loadingEllipsis steps(4,end) 2s infinite;
  animation: loadingEllipsis steps(4,end) 2s infinite;
}

.c5::before {
  color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
}

.c0 > nav {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 > main {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

@media print {
  .c1 {
    display: none;
  }
}

@media print {
  .c2 {
    margin: 0;
    max-width: 100%;
  }
}

@media (min-width:480px) {
  .c3 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c4 {
    line-height: 1.3;
  }
}

@media print {
  .c0 {
    display: block;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
        id="audiofocus"
      >
        <h1
          aria-label="Printing Official Ballot."
        >
          <div
            class="c4"
          >
            <h1
              aria-label="Printing Official Ballot."
              class="c5"
            >
              Printing Official Ballot
            </h1>
          </div>
        </h1>
      </div>
    </div>
  </main>
</div>
`;

exports[`renders PrintPage without votes and inline seal 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c2 {
  margin: auto 0;
}

.c3 {
  margin: auto;
  max-width: 66ch;
  text-align: center;
  line-height: 1.2;
}

.c3 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c3 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c3 h3 {
  font-size: 1.17rem;
}

.c3 h4 {
  font-size: 1rem;
}

.c3 h5 {
  font-size: 0.9rem;
}

.c3 h3,
.c3 h4,
.c3 h5,
.c3 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p {
  margin-top: -0.75rem;
}

.c3 h3 + p,
.c3 h4 + p,
.c3 h5 + p {
  margin-top: -1rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c3 dl {
  margin: 1rem 0;
}

.c4 {
  max-width: 66ch;
  line-height: 1.2;
}

.c4 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c4 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c4 h3 {
  font-size: 1.17rem;
}

.c4 h4 {
  font-size: 1rem;
}

.c4 h5 {
  font-size: 0.9rem;
}

.c4 h3,
.c4 h4,
.c4 h5,
.c4 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c4 h1 + h2 {
  margin-top: -0.75rem;
}

.c4 h1 + p,
.c4 h2 + p {
  margin-top: -0.75rem;
}

.c4 h3 + p,
.c4 h4 + p,
.c4 h5 + p {
  margin-top: -1rem;
}

.c4 > :first-child {
  margin-top: 0;
}

.c4 > :last-child {
  margin-bottom: 0;
}

.c4 dl {
  margin: 1rem 0;
}

.c5 {
  margin-left: -1em;
  text-align: center;
  white-space: nowrap;
}

.c5::before,
.c5::after {
  display: inline-block;
  width: 0;
  overflow: hidden;
  vertical-align: bottom;
  text-align: left;
  content: '…';
  -webkit-animation: loadingEllipsis steps(4,end) 2s infinite;
  animation: loadingEllipsis steps(4,end) 2s infinite;
}

.c5::before {
  color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
}

.c0 > nav {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 > main {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

@media print {
  .c1 {
    display: none;
  }
}

@media print {
  .c2 {
    margin: 0;
    max-width: 100%;
  }
}

@media (min-width:480px) {
  .c3 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c4 {
    line-height: 1.3;
  }
}

@media print {
  .c0 {
    display: block;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
        id="audiofocus"
      >
        <h1
          aria-label="Printing Official Ballot."
        >
          <div
            class="c4"
          >
            <h1
              aria-label="Printing Official Ballot."
              class="c5"
            >
              Printing Official Ballot
            </h1>
          </div>
        </h1>
      </div>
    </div>
  </main>
</div>
`;

exports[`renders PrintPage without votes and no seal 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c2 {
  margin: auto 0;
}

.c3 {
  margin: auto;
  max-width: 66ch;
  text-align: center;
  line-height: 1.2;
}

.c3 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c3 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c3 h3 {
  font-size: 1.17rem;
}

.c3 h4 {
  font-size: 1rem;
}

.c3 h5 {
  font-size: 0.9rem;
}

.c3 h3,
.c3 h4,
.c3 h5,
.c3 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p {
  margin-top: -0.75rem;
}

.c3 h3 + p,
.c3 h4 + p,
.c3 h5 + p {
  margin-top: -1rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c3 dl {
  margin: 1rem 0;
}

.c4 {
  max-width: 66ch;
  line-height: 1.2;
}

.c4 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c4 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c4 h3 {
  font-size: 1.17rem;
}

.c4 h4 {
  font-size: 1rem;
}

.c4 h5 {
  font-size: 0.9rem;
}

.c4 h3,
.c4 h4,
.c4 h5,
.c4 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c4 h1 + h2 {
  margin-top: -0.75rem;
}

.c4 h1 + p,
.c4 h2 + p {
  margin-top: -0.75rem;
}

.c4 h3 + p,
.c4 h4 + p,
.c4 h5 + p {
  margin-top: -1rem;
}

.c4 > :first-child {
  margin-top: 0;
}

.c4 > :last-child {
  margin-bottom: 0;
}

.c4 dl {
  margin: 1rem 0;
}

.c5 {
  margin-left: -1em;
  text-align: center;
  white-space: nowrap;
}

.c5::before,
.c5::after {
  display: inline-block;
  width: 0;
  overflow: hidden;
  vertical-align: bottom;
  text-align: left;
  content: '…';
  -webkit-animation: loadingEllipsis steps(4,end) 2s infinite;
  animation: loadingEllipsis steps(4,end) 2s infinite;
}

.c5::before {
  color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
}

.c0 > nav {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 > main {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

@media print {
  .c1 {
    display: none;
  }
}

@media print {
  .c2 {
    margin: 0;
    max-width: 100%;
  }
}

@media (min-width:480px) {
  .c3 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c4 {
    line-height: 1.3;
  }
}

@media print {
  .c0 {
    display: block;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
        id="audiofocus"
      >
        <h1
          aria-label="Printing Official Ballot."
        >
          <div
            class="c4"
          >
            <h1
              aria-label="Printing Official Ballot."
              class="c5"
            >
              Printing Official Ballot
            </h1>
          </div>
        </h1>
      </div>
    </div>
  </main>
</div>
`;
